File: leetcode/BackTracking/leetcode-0017/17.go
package backtracking

func letterCombinations(digits string) []string {
	res := []string{}

	if len(digits) == 0 {
		return res
	}

	letters := []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	helper(&res, letters, "", digits, 0)
	return res
}

func helper(res *[]string, letters []string, str, digits string, i int) {
	if i == len(digits) {
		*res = append(*res, str)
		return
	}

	letter := letters[digits[i]-'0']
	for j := 0; j < len(letter); j++ {
		helper(res, letters, str+string(letter[j]), digits, i+1)
	}
}

File: leetcode/BackTracking/leetcode-0017/17.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return res;
        }

        String[] letters = new String[]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        helper(res, letters, "", digits, 0);
        return res;
    }

    private void helper(List<String> res, String[] letters, String str, String digits, int i) {
        if (i == digits.length()) {
            res.add(str);
            return;
        }

        String letter = letters[digits.charAt(i) - '0'];
        for (int j = 0; j < letter.length(); j++) {
            helper(res, letters, str + letter.charAt(j), digits, i + 1);
        }
    }
}